pkg/scrape/discovery/consul/consul.go

```diff
@@ -0,0 +1,583 @@
+	serviceLabel = model.MetaLabelPrefix + "consul_service"
+	// healthLabel is the name of the label containing the health of the service instance
+	healthLabel = model.MetaLabelPrefix + "consul_health"
+	// serviceAddressLabel is the name of the label containing the (optional) service address.
+	serviceAddressLabel = model.MetaLabelPrefix + "consul_service_address"
+	// servicePortLabel is the name of the label containing the service port.
+	servicePortLabel = model.MetaLabelPrefix + "consul_service_port"
+	// datacenterLabel is the name of the label containing the datacenter ID.
+	datacenterLabel = model.MetaLabelPrefix + "consul_dc"
+	// namespaceLabel is the name of the label containing the namespace (Consul Enterprise only).
+	namespaceLabel = model.MetaLabelPrefix + "consul_namespace"
+	// partitionLabel is the name of the label containing the Admin Partition (Consul Enterprise only).
+	partitionLabel = model.MetaLabelPrefix + "consul_partition"
+	// taggedAddressesLabel is the prefix for the labels mapping to a target's tagged addresses.
+	taggedAddressesLabel = model.MetaLabelPrefix + "consul_tagged_address_"
+	// serviceIDLabel is the name of the label containing the service ID.
+	serviceIDLabel = model.MetaLabelPrefix + "consul_service_id"
+
+	// Constants for instrumentation.
+	namespace = "pyroscope"
+)
+
+var (
+	rpcFailuresCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: namespace,
+			Name:      "sd_consul_rpc_failures_total",
+			Help:      "The number of Consul RPC call failures.",
+		})
+	rpcDuration = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace:  namespace,
+			Name:       "sd_consul_rpc_duration_seconds",
+			Help:       "The duration of a Consul RPC call in seconds.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{"endpoint", "call"},
+	)
+
+	// Initialize metric vectors.
+	servicesRPCDuration = rpcDuration.WithLabelValues("catalog", "services")
+	serviceRPCDuration  = rpcDuration.WithLabelValues("catalog", "service")
+
+	// DefaultSDConfig is the default Consul SD configuration.
+	DefaultSDConfig = SDConfig{
+		TagSeparator:     ",",
+		Scheme:           "http",
+		Server:           "localhost:8500",
+		AllowStale:       true,
+		RefreshInterval:  model.Duration(30 * time.Second),
+		HTTPClientConfig: config.DefaultHTTPClientConfig,
+	}
+)
+
+func init() {
+	discovery.RegisterConfig(&SDConfig{})
+	prometheus.MustRegister(rpcFailuresCount, rpcDuration)
+}
+
+// SDConfig is the configuration for Consul service discovery.
+type SDConfig struct {
+	Server       string        `yaml:"server,omitempty"`
+	Token        config.Secret `yaml:"token,omitempty"`
+	Datacenter   string        `yaml:"datacenter,omitempty"`
+	Namespace    string        `yaml:"namespace,omitempty"`
+	Partition    string        `yaml:"partition,omitempty"`
+	TagSeparator string        `yaml:"tag_separator,omitempty"`
```

Comment:
There is one annoying difference with the Prometheus configuration, which is that in Pyroscope we use `kebab-style` (e.g `yaml:"tag-separator,omitempty"`). Could you please change yaml keys accordingly to ensure the format is consistent

Reply:
Fixed~